### ISSUE TITLE
There should be wrong when transfer virtualservice resource to envoy config 

### ISSUE THREAD

**Issue by haifzhu** [istio/istio#51232]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

when configure the virtualservice resource
~~~
apiVersion: v1
items:
- apiVersion: networking.istio.io/v1beta1
  kind: VirtualService
  metadata:
    labels:
      common: "true"
    name: common.zcy-gateway
  spec:
    gateways:
    - zcy-gateway
    hosts:
    - supplier-staging.a.com
    - vehicle-staging.a.com
    http:
    - match:
      - uri:
          prefix: /api/good/compose/suit
      - uri:
          prefix: /api/agreement-center/goods/standard
      - uri:
          prefix: /api/agreement-center/brandItem/
      - uri:
          prefix: /agreement/goods/channelGoods/
      - uri:
          prefix: /agreement/goods/agreementItem/
      - uri:
          exact: /agreement/goods/channel/paging
      - uri:
          exact: /agreement/goods/protocol/paging
      - uri:
          exact: /agreement/goods/protocol/paging/v2
      - uri:
          exact: /agreement/goods/findChannelTypes
      - uri:
          exact: /agreement/goods/findParts
      - uri:
          exact: /agreement/goods/findPartsTag
      - uri:
          exact: /agreement/goods/getAuditTabs
      - uri:
          exact: /agreement/goods/pagingToAudit
      - uri:
          exact: /agreement/goods/pagingToSumAudit
      - uri:
          exact: /agreement/goods/pagingAudited
      - uri:
          exact: /agreement/goods/deleteProtocolGood
      - uri:
          exact: /agreement/goods/addProtocolGoods
      - uri:
          exact: /agreement/goods/his
      - uri:
          exact: /agreement/goods/pagingProtocolGoodsByBid
      - uri:
          exact: /agreement/goods/pageProtocolGoodsByBid
      - uri:
          exact: /agreement/goods/checkGoodsDiscountRate
      - uri:
          exact: /agreement/goods/snapshot
      - uri:
          exact: /agreement/goods/getProtocolUnTagGoodsWrongMsg
      - uri:
          exact: /agreement/goods/mark
      - uri:
          exact: /agreement/goods/canBeAuthorized
      - uri:
          exact: /agreement/goods/brandItemList
      - uri:
          exact: /agreement/goods/export/agGoods
      - uri:
          exact: /agreement/goods/getItemFeatures
      - uri:
          exact: /agreement/goods/getProtocolGoodParamOffsetData
      - uri:
          exact: /agreement/goods/getNextAudit
      - uri:
          exact: /agreement/goods/mergeByShop
      - uri:
          exact: /agreement/goods/getFlowKey
      - uri:
          exact: /agreement/goods/getSupplierGoodServices
      - uri:
          exact: /agreement/goods/judgeWhetherToMerge
      - uri:
          exact: /agreement/goods/judgeShopStartMerge
      - uri:
          exact: /agreement/goods/api/micro/itemUpgrade/upgradeCount
      - uri:
          exact: /agreement/goods/api/micro/itemUpgrade/remindFlag
      - uri:
          exact: /agreement/goods/admin/pagingToAudit
      - uri:
          prefix: /api/goods
      name: web-agreement.skh5Z
      route:
      - destination:
          host: web-aggregated.preliminary.svc.cluster.local
          port:
            number: 8080
- apiVersion: networking.istio.io/v1beta1
  kind: VirtualService
  metadata:
    labels:
      default: "true"
    name: supplier-staging.zcy-gateway.default
  spec:
    gateways:
    - zcy-gateway
    hosts:
    - supplier-staging.a.com
    http:
    - route:
      - destination:
          host: web-supplier.preliminary.svc.cluster.local
          port:
            number: 8080
- apiVersion: networking.istio.io/v1beta1
  kind: VirtualService
  metadata:
    labels:
      default: "true"
    name: vehicle-staging.zcy-gateway.default
  spec:
    gateways:
    - zcy-gateway
    hosts:
    - vehicle-staging.a.com
    http:
    - match:
      - uri:
          prefix: /
      route:
      - destination:
          host: web-car.preliminary.svc.cluster.local
          port:
            number: 8080
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
~~~

Use istioctl command dump envoy config 

~~~
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /api/good/compose/suit*                                 common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /api/agreement-center/goods/standard*                   common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /api/agreement-center/brandItem/*                       common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/channelGoods/*                         common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/agreementItem/*                        common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/channel/paging                         common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/protocol/paging                        common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/protocol/paging/v2                     common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/findChannelTypes                       common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/findParts                              common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/findPartsTag                           common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/getAuditTabs                           common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/pagingToAudit                          common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/pagingToSumAudit                       common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/pagingAudited                          common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/deleteProtocolGood                     common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/addProtocolGoods                       common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/his                                    common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/pagingProtocolGoodsByBid               common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/pageProtocolGoodsByBid                 common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/checkGoodsDiscountRate                 common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/snapshot                               common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/getProtocolUnTagGoodsWrongMsg          common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/mark                                   common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/canBeAuthorized                        common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/brandItemList                          common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/export/agGoods                         common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/getItemFeatures                        common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/getProtocolGoodParamOffsetData         common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/getNextAudit                           common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/mergeByShop                            common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/getFlowKey                             common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/getSupplierGoodServices                common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/judgeWhetherToMerge                    common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/judgeShopStartMerge                    common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/api/micro/itemUpgrade/upgradeCount     common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/api/micro/itemUpgrade/remindFlag       common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /agreement/goods/admin/pagingToAudit                    common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /api/goods*                                             common.zcy-gateway.staging
http.8080     supplier-staging.a.com:80                       supplier-staging.a.com, vehicle-staging.a.com     /*                                                      vehicle-staging.zcy-gateway.default.staging
~~~

The default route config for supplier-staging.a.com is wrong, there should be supplier-staging.zcy-gateway.default.staging, but vehicle-staging.zcy-gateway.default.staging .

### Version

```Text
# istioctl version
client version: 1.19.3
control plane version: 1.16.3
data plane version: 1.16.3 (1 proxies)
# kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.17", GitCommit:"953be8927218ec8067e1af2641e540238ffd7576", GitTreeState:"clean", BuildDate:"2023-02-22T13:34:27Z", GoVersion:"go1.19.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.17", GitCommit:"953be8927218ec8067e1af2641e540238ffd7576", GitTreeState:"clean", BuildDate:"2023-02-22T13:27:46Z", GoVersion:"go1.19.6", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by howardjohn** [comment]
Istio 1.16 is LONG since EOL. I don't see this happening on newer versions. Please upgrade to a supported version and re-open if you still see the issue.

**Comment by haifzhu** [comment]
> Istio 1.16 is LONG since EOL. I don't see this happening on newer versions. Please upgrade to a supported version and re-open if you still see the issue.

After test, the issue is fixed on  istio newest version 1.22.0, but still exist on istio version 1.21.2, 1.20.6,1.19.10。

**Comment by haifzhu** [comment]
the issue is fixed on commit 0cb5c33595cdfaea732178a4d70265ac0a762255
~~~
commit 0cb5c33595cdfaea732178a4d70265ac0a762255
Author: Hannes Östlund <hannes.ostlund@gmail.com>
Date:   Wed Apr 10 20:27:40 2024 +0200

    Fix bug when virtual services overwrites routes (#50343)

    * Fix bug when virtual services overwrites routes

    Make sure that each host has its own slice of routes to not
    overwrite routes added by other virtual services.

    * Update pilot/pkg/networking/core/gateway.go

    Co-authored-by: John Howard <johnbhoward96@gmail.com>
~~~
